benchmarks/Microsoft.AspNetCore.Mvc.Performance/MvcEndpointDatasourceBenchmark.cs

```diff
@@ -0,0 +1,138 @@
+
+        private MockActionDescriptorCollectionProvider _conventionalActionDescriptorCollectionProvider;
+        private MockActionDescriptorCollectionProvider _attributeActionDescriptorCollectionProvider;
+        private List<MvcEndpointInfo> _conventionalEndpointInfos;
+
+        [Params(1, 100, 1000)]
+        public int ActionCount;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            _conventionalActionDescriptorCollectionProvider = new MockActionDescriptorCollectionProvider(
+                Enumerable.Range(0, ActionCount).Select(i => CreateActionDescriptor(i, false)).ToList()
+                );
+
+            _attributeActionDescriptorCollectionProvider = new MockActionDescriptorCollectionProvider(
+                Enumerable.Range(0, ActionCount).Select(i => CreateActionDescriptor(i, true)).ToList()
+                );
+
+            _conventionalEndpointInfos = new List<MvcEndpointInfo>
+            {
+                new MvcEndpointInfo(
+                    "Default",
+                    DefaultRoute,
+                    new RouteValueDictionary(),
+                    new Dictionary<string, object>(),
+                    new RouteValueDictionary(),
+                    new MockInlineConstraintResolver())
+            };
+        }
+
+        [Benchmark]
+        public void AttributeRouteEndpoints()
+        {
+            var endpointDataSource = CreateMvcEndpointDataSource(_attributeActionDescriptorCollectionProvider);
+            var endpoints = endpointDataSource.Endpoints;
+        }
+
+        [Benchmark]
+        public void ConventionalEndpoints()
+        {
+            var endpointDataSource = CreateMvcEndpointDataSource(_conventionalActionDescriptorCollectionProvider);
+            endpointDataSource.ConventionalEndpointInfos.AddRange(_conventionalEndpointInfos);
+            var endpoints = endpointDataSource.Endpoints;
+        }
+
+        private ActionDescriptor CreateActionDescriptor(int id, bool attributeRoute)
+        {
+            var actionDescriptor = new ActionDescriptor
+            {
+                RouteValues = new Dictionary<string, string>
+                {
+                    ["Controller"] = "Controller" + id,
+                    ["Action"] = "Index"
+                },
+                DisplayName = "Action " + id
+            };
+
+            if (attributeRoute)
+            {
+                actionDescriptor.AttributeRouteInfo = new AttributeRouteInfo
+                {
+                    Template = DefaultRoute
+                };
+            }
+
+            return actionDescriptor;
+        }
+
+        private MvcEndpointDataSource CreateMvcEndpointDataSource(
+            IActionDescriptorCollectionProvider actionDescriptorCollectionProvider = null,
+            MvcEndpointInvokerFactory mvcEndpointInvokerFactory = null,
+            IEnumerable<IActionDescriptorChangeProvider> actionDescriptorChangeProviders = null)
```

Comment:
It looks like just the first parameter is used here? will you ever need the other ones?

Reply:
ok! I just copied and pasted it from unit tests